### ISSUE TITLE
lsp-keymap-prefix in evil modes

### ISSUE THREAD

**Issue by rdbeni0** [emacs-lsp/lsp-mode#1530]
hi

i am trying to configure \ as prefix key
`(setq lsp-keymap-prefix "\\")` but only in evil states like: normal and visual

basically there is general.el package : https://github.com/noctuid/general.el
and have this in my init.el:

```
(general-define-key  
   :prefix "\\"  
   :states '(normal visual emacs)  
   :keymaps 'override  
     "=" '(:ignore t :which-key "format")  
     "=b" '(lsp-format-buffer :which-key "format-buffer")  
     "=r" '(lsp-format-region :which-key "format-region")  
     "g" '(:ignore t :which-key "goto")  
     "gt" '(lsp-goto-type-definition :which-key "goto-type-definition")  
     "T" '(:ignore t :which-key "Toggle")  
     "r" '(:ignore t :which-key "refactor")  
     "rr" '(lsp-rename :which-key "rename")  
)  
```
^ and this working as expected, but i need to insert each keybinding manually.
However i don't want manually insert all keybindings, it will be better just use lsp-keymap-prefix in correct evil modes.

so how can i configure lsp-keymap-prefix , but only in some evil modes?

**Comment by yyoncho** [comment]
Don't use lsp-keymap-prefix but bind lsp-command-map using general.

**Comment by yyoncho** [comment]
FTR here is is an example:

```
(evil-define-key 'normal lsp-mode-map (kbd "\\") lsp-command-map)
```

**Comment by sflomenb** [comment]
One issue I am having with this is that when I open a file, LSP mode starts, and I input my prefix, it says it is undefined. If I go into insert mode and back to normal mode, then the prefix works.

Not sure if this is an lsp-mode issue or an evil-mode issue. However, I have defined other commands using evil-define-key in normal mode and this does not happen with those.

**Comment by meliache** [comment]
I had the same issue as @sflomenb . Seems it's caused by https://github.com/emacs-evil/evil/issues/301, where some workarounds are posted. I solved it by using `evil-define-minor-mode-key`
``` emacs-lisp
  (evil-define-minor-mode-key 'normal lsp-mode (kbd "SPC l") lsp-command-map)
```
or the `general.el` wrapper:
``` emacs-lisp
(general-def 'normal lsp-mode :definer 'minor-mode
      "SPC l" lsp-command-map)
```
Another solution would be to run `evil-normal-keymaps` in a hook.

**Comment by programmador** [comment]
Here's my example. Though I'm using elpaca so syntax may differ for other package installation options.
Nevertheless I post it almost unchanged, just cut few unnecessary variable sets.

```
(use-package lsp-mode
  :after (which-key)
  :ensure t
  :config
  (setq lsp-keymap-prefix ",")
  (evil-define-key 'normal lsp-mode-map (kbd ",") lsp-command-map)
  :hook
  (php-mode . lsp-deferred)
  (lsp-mode . evil-normalize-keymaps)
  (lsp-mode . lsp-enable-which-key-integration)
  :commands (lsp lsp-deferred))
```

So that's what is interesting here and why I decided to post it hopefully to save someone's time:

- `evil-define-minor-mode-key` didn't work for me so I've called `evil-normalize-keymaps` via hook
- `setq lsp-keymap-prefix` is called anyway. Without doing it `lsp-enable-which-key-integration` will not work
